### PR TITLE
Reject uneven splits in Chunk weight conversions

### LINKED ISSUE #48323: Chunk conversion op silently produces unequal chunks when the fused tensor is not divisible by the target count
### System Info

- transformers 5.16.0.dev0 (current main)
- python 3.12.6, Windows 11 (not platform specific)

### Reproduction

The v5 dynamic loading API's `Chunk` op (`src/transformers/core_model_loading.py`, class `Chunk`) splits a fused tensor for targets like q/k/v. Its guard only compares chunk COUNT against the target count:

```python
chunks = tuple(chunk.contiguous() for chunk in torch.chunk(tensor, sizes, dim=self.dim))
if len(input_dict) > 1 or len(target_patterns) == 1 or len(chunks) != len(target_patterns):
    raise ValueError(f"Failed to convert {kwargs.get('full_layer_name')}")
```

`torch.chunk` returns unequal chunks when the tensor size is not divisible by the requested count, and can still return exactly as many chunks as there are targets. Then every downstream consumer receives differently sized tensors:

```python
import torch
from transformers.core_model_loading import Chunk

op = Chunk(dim=0)
t = {"x": [torch.arange(5.)]}
out = op.convert(t, source_patterns=["x"], target_patterns=["a", "b"], full_layer_name="demo.layer")
print({k: tuple(v.shape) for k, v in out.items()})
```

Output on current main - no error, wrong shapes:

```text
{'a': (3,), 'b': (2,)}
```

In a real `from_pretrained` conversion this surfaces either as a confusing per-target shape mismatch error far from the cause, or worse: silently re-initializing weights under `ignore_mismatched_sizes`, which corrupts the model without any signal.

### Expected behavior

Since the op's own contract is "Split a tensor along dim into equally sized chunks", an indivisible split should fail loudly at the conversion site with a clear message naming the violated contract and the layer, instead of producing silently unequal chunks. Every registered `Chunk` usage in the codebase splits equal-sized siblings by construction, so a divisibility check would not reject any legitimate existing use.

Duplicate check: searched open/closed issues and PRs around torch.chunk uneven splits, Chunk.convert, and ignore_mismatched_sizes silent re-init; nothing tracks this.

### PR DESCRIPTION
**Pull request by Ki-Seki**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48405) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48405)

# What does this PR do?

This PR fixes the issue where the `Chunk` weight conversion operation silently generates unequal-length slices when the tensor cannot be evenly divided by the target number.

Fixes #48323

## Code Agent Policy

- [ ] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

Not applicable: I have previously contributed to Transformers.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@Cyrilvallez
@ArthurZucker
@VaggelisGian

### CODE REVIEW

**Comment by Ki-Seki** [comment]
Below is a real-world example:

```python
import torch

from transformers import Qwen4ExpTextConfig
from transformers.core_model_loading import Chunk


config = Qwen4ExpTextConfig(split_ngram_parts=3)
operation = Chunk(
    dim=0,
    num_shards_attribute="split_ngram_parts",
)

output = operation.convert(
    {
        "ngram_embedding.weight": [
            torch.zeros(5, 4),
        ]
    },
    source_patterns=["ngram_embedding.weight"],
    target_patterns=["ngram_embedding.shard_*.weight"],
    config=config,
    full_layer_name=(
        "model.layers.1.ple.ple_embedding."
        "ngram_embedding.weight"
    ),
)

print({name: tuple(tensor.shape) for name, tensor in output.items()})
```

**Comment by Cyrilvallez** [comment]
Theoretical AI fix
